## Упражнения по библиотеке Numpy

In [262]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [263]:
# python solution
import random

array_size = random.randint(1, 10)
arr = [random.randint(0, 10) for _ in range(array_size)]
for index, value in enumerate(arr):
    if 3 < value < 8:
        arr[index] = -value
arr

[3, 2, -4, 1, 8, 8, -7, -6, 9, 10]

In [264]:
# numpy solution
arr = np.random.randint(0, 10, size=np.random.randint(1, 15)) # size and range are random
mask = (arr > 3) & (arr < 8)
arr[mask] = -arr[mask]
arr

array([-4,  8,  3])

**2.** Заменить максимальный элемент случайного массива на 0

In [265]:
# python solution
import random

array_size = random.randint(1, 10)
arr = [random.randint(0, 10) for _ in range(array_size)]
max_value = max(arr)
for index, value in enumerate(arr):
    if value == max_value:
        arr[index] = 0 
arr

[8, 7, 7, 0, 7, 6]

In [266]:
# numpy solution
arr = np.random.randint(0, 10, size=np.random.randint(1, 15)) # size and range are random
max_value = arr.max()
arr[arr == max_value] = 0
arr

array([0, 0])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [267]:
# python solution
from itertools import product
import random

def straight_product_of_arrays(array):
    return list(product(*array))

# example
lists = [[random.randint(0, 10) for _ in range(random.randint(2, 6))] 
         for _ in range(random.randint(2, 5))]
input_array = lists 

result = straight_product_of_arrays(input_array)
print(input_array)
print(result)


[[10, 3, 3, 0, 0], [0, 2, 10]]
[(10, 0), (10, 2), (10, 10), (3, 0), (3, 2), (3, 10), (3, 0), (3, 2), (3, 10), (0, 0), (0, 2), (0, 10), (0, 0), (0, 2), (0, 10)]


In [268]:
# numpy solution

from itertools import product

def straight_product_of_arrays(array: np.array) -> np.ndarray:
    return np.array(list(product(*array)))

# example
lists = [np.random.randint(0, 10, size=np.random.randint(2, 6)) for _ in range(np.random.randint(2, 5))]
input_array = np.array(lists, dtype=object)

result = straight_product_of_arrays(input_array)
print(input_array)
print(result)

[array([5, 0, 1, 5]) array([4, 5, 0, 8, 8]) array([8, 8, 7]) array([6, 3])]
[[5 4 8 6]
 [5 4 8 3]
 [5 4 8 6]
 [5 4 8 3]
 [5 4 7 6]
 [5 4 7 3]
 [5 5 8 6]
 [5 5 8 3]
 [5 5 8 6]
 [5 5 8 3]
 [5 5 7 6]
 [5 5 7 3]
 [5 0 8 6]
 [5 0 8 3]
 [5 0 8 6]
 [5 0 8 3]
 [5 0 7 6]
 [5 0 7 3]
 [5 8 8 6]
 [5 8 8 3]
 [5 8 8 6]
 [5 8 8 3]
 [5 8 7 6]
 [5 8 7 3]
 [5 8 8 6]
 [5 8 8 3]
 [5 8 8 6]
 [5 8 8 3]
 [5 8 7 6]
 [5 8 7 3]
 [0 4 8 6]
 [0 4 8 3]
 [0 4 8 6]
 [0 4 8 3]
 [0 4 7 6]
 [0 4 7 3]
 [0 5 8 6]
 [0 5 8 3]
 [0 5 8 6]
 [0 5 8 3]
 [0 5 7 6]
 [0 5 7 3]
 [0 0 8 6]
 [0 0 8 3]
 [0 0 8 6]
 [0 0 8 3]
 [0 0 7 6]
 [0 0 7 3]
 [0 8 8 6]
 [0 8 8 3]
 [0 8 8 6]
 [0 8 8 3]
 [0 8 7 6]
 [0 8 7 3]
 [0 8 8 6]
 [0 8 8 3]
 [0 8 8 6]
 [0 8 8 3]
 [0 8 7 6]
 [0 8 7 3]
 [1 4 8 6]
 [1 4 8 3]
 [1 4 8 6]
 [1 4 8 3]
 [1 4 7 6]
 [1 4 7 3]
 [1 5 8 6]
 [1 5 8 3]
 [1 5 8 6]
 [1 5 8 3]
 [1 5 7 6]
 [1 5 7 3]
 [1 0 8 6]
 [1 0 8 3]
 [1 0 8 6]
 [1 0 8 3]
 [1 0 7 6]
 [1 0 7 3]
 [1 8 8 6]
 [1 8 8 3]
 [1 8 8 6]
 [1 8 8 3]
 [1 8 7 6]
 [1 8 7 3]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [269]:
# python solution

import random

array_a = [[random.randint(1, 20) for _ in range(3)] for _ in range(8)]
array_b = [[random.randint(1, 20) for _ in range(2)] for _ in range(2)]

result = []

for i, row_a in enumerate(array_a):
    first_b_row = False
    second_b_row = False
    for elem in row_a:
        if elem in array_b[0]:
            first_b_row = True
            break
    for elem in row_a:
        if elem in array_b[1]:
            second_b_row = True
            break
    if first_b_row and second_b_row:
        result.append(row_a)
    

array_a, array_b, result

([[17, 6, 8],
  [9, 16, 17],
  [11, 9, 7],
  [5, 9, 11],
  [14, 18, 16],
  [20, 18, 15],
  [20, 13, 9],
  [1, 11, 8]],
 [[20, 20], [15, 3]],
 [[20, 18, 15]])

In [270]:
# numpy solution

array_a = np.random.randint(1, 20, size=(8, 3))
array_b = np.random.randint(1, 20, size=(2, 2))

b0 = set(array_b[0])
b1 = set(array_b[1])

mask_b0 = np.isin(array_a, list(b0)).any(axis=1)
mask_b1 = np.isin(array_a, list(b1)).any(axis=1)

final_mask = mask_b0 & mask_b1

result = array_a[final_mask]

array_a, array_b, result

(array([[ 3, 19, 11],
        [ 7, 10,  8],
        [16, 13,  9],
        [19,  9,  6],
        [13,  1,  7],
        [10,  8,  4],
        [11, 19, 17],
        [17, 17, 18]]),
 array([[11, 18],
        [ 3,  3]]),
 array([[ 3, 19, 11]]))

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [271]:
# python solution

import random

array = [[random.randint(1, 5) for _ in range(3)] for _ in range(10)]
filtered_array = []
for row in array:
    if len(set(row)) > 1:
        filtered_array.append(row)

array, filtered_array

([[2, 4, 5],
  [2, 2, 4],
  [1, 3, 1],
  [5, 2, 4],
  [5, 3, 3],
  [5, 4, 5],
  [2, 2, 4],
  [4, 3, 5],
  [1, 1, 3],
  [1, 5, 1]],
 [[2, 4, 5],
  [2, 2, 4],
  [1, 3, 1],
  [5, 2, 4],
  [5, 3, 3],
  [5, 4, 5],
  [2, 2, 4],
  [4, 3, 5],
  [1, 1, 3],
  [1, 5, 1]])

In [272]:
# numpy solution

array = np.random.randint(1, 6, size=(10, 3))

filtered_array = array[np.apply_along_axis(lambda row: len(np.unique(row)) > 1, axis=1, arr=array)]

array, filtered_array

(array([[2, 3, 1],
        [4, 3, 5],
        [1, 2, 4],
        [1, 1, 3],
        [2, 1, 3],
        [2, 1, 3],
        [3, 5, 5],
        [4, 1, 4],
        [3, 1, 4],
        [2, 1, 5]]),
 array([[2, 3, 1],
        [4, 3, 5],
        [1, 2, 4],
        [1, 1, 3],
        [2, 1, 3],
        [2, 1, 3],
        [3, 5, 5],
        [4, 1, 4],
        [3, 1, 4],
        [2, 1, 5]]))

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [273]:
# python solution

array = [[random.randint(1, 3) for _ in range(2)] for _ in range(5)]

unique_array = []

for row in array:
    if row not in unique_array:
        unique_array.append(row)

array, unique_array

([[1, 3], [2, 1], [1, 1], [1, 1], [1, 1]], [[1, 3], [2, 1], [1, 1]])

In [274]:
# numpy solution

array = np.random.randint(1, 4, size=(5, 2))
unique_array = np.unique(array, axis=0)

array, unique_array

(array([[3, 1],
        [2, 3],
        [3, 2],
        [2, 1],
        [2, 3]]),
 array([[2, 1],
        [2, 3],
        [3, 1],
        [3, 2]]))

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [ ]:
def product_of_diagonal_elements(matrix):
    diagonal_elements = np.diagonal(matrix)
    nonzero_elements_mask = diagonal_elements[diagonal_elements != 0]
    return np.prod(nonzero_elements_mask)

X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
result = product_of_diagonal_elements(X)
print(result)

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [ ]:
# multiset is a set that contains similar elems and every elem is found the same amount
def multiset_check(x, y):
    x_vals, x_counts = np.unique(x, return_counts=True)
    y_vals, y_counts = np.unique(y, return_counts=True)
    return np.array_equal(x_vals, y_vals) and np.array_equal(x_counts, y_counts)

x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
multiset_check(x, y)

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [ ]:
x = np.array([10, 2, 0, 8, 0, 0, 5, 7, 0])

mask = np.zeros_like(x, dtype=bool)
mask[1:] = (x[:-1] == 0)  
elements_after_zero = x[mask]

max_after_zero = np.max(elements_after_zero) if len(elements_after_zero) > 0 else None

print(max_after_zero)

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [ ]:
def run_length_encoding(x):
    if len(x) == 0:
        return np.array([]), np.array([])
    
    change_indices = np.where(np.diff(x) != 0)[0] + 1
    indices = np.concatenate(([0], change_indices, [len(x)]))
    values = x[indices[:-1]]
    counts = np.diff(indices)
    return values, counts

x = np.array([2, 2, 2, 3, 3, 3, 5])
values, counts = run_length_encoding(x)
values, counts

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [ ]:
import time
from scipy.spatial.distance import cdist

np.random.seed(0) # for fixing results
X = np.random.rand(1000, 50)
Y = np.random.rand(1000, 50)

def euclidean_distance_matrix_manual(X, Y):
    X_sq = np.sum(X**2, axis=1).reshape(-1, 1)  
    Y_sq = np.sum(Y**2, axis=1).reshape(1, -1)  
    XY = X @ Y.T                                
    dists = np.sqrt(X_sq - 2 * XY + Y_sq)
    return dists

start_manual = time.time()
D_manual = euclidean_distance_matrix_manual(X, Y)
end_manual = time.time()

start_scipy = time.time()
D_scipy = cdist(X, Y, metric='euclidean')
end_scipy = time.time()

np.abs(D_manual - D_scipy).max()

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [280]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [281]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [305]:
crunchie_munchies_calories = 60
average_calories = calorie_stats.mean()
diff = average_calories - crunchie_munchies_calories
average_calories, diff


(np.float64(106.88311688311688), np.float64(46.883116883116884))

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [312]:
calorie_stats_sorted = np.sort(calorie_stats)

calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [314]:
median_calories = np.median(calorie_stats_sorted)
median_calories, average_calories

(np.float64(110.0), np.float64(106.88311688311688))

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [317]:
nth_percintile = None
for percintile in range(0, 101):
    current = np.percentile(calorie_stats_sorted, percintile)
    print(current)
    if current > 60:
        nth_percintile = current
        break

nth_percintile

50.0
50.0
50.0
55.599999999999994
70.0


np.float64(70.0)

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [318]:
more_calories = np.sum(calorie_stats > 60) / len(calorie_stats)
more_calories

np.float64(0.961038961038961)

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [319]:
calorie_std = np.std(calorie_stats_sorted)
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

В интересах компании Mycrunch можно эффективно использовать значение more_calories, предоставляя продукт как из раздела для здорового питания, показывая, что его энергетическая ценность лучше для этой категории, чем у 96% хлопьев на рынке. Также можно удачно использовать значение медианы, потому что для продвижения товара хорошо показывать, какия медианные значения по рынку данного товара.